solutions/usaco-553.mdx

```diff
@@ -0,0 +1,97 @@
+---
+id: usaco-553
+source: USACO Gold 2015 Open
+title: Palindromic Paths
+author: Maggie Liu
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_palpath_gold.html)
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N^3)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <cstdio>
+#include <string>
+#include <algorithm>
+using namespace std;
+
+int main()
+{
+	freopen("palpath.in", "r", stdin);
+	freopen("palpath.out", "w", stdout);
+	int n;
+	cin >> n;
+	string grid[500];
+	for (int i = 0; i < n; i++)
+	{
+		cin >> grid[i];
+	}
+	long long prev[500][500]{};
+	for (int i = 0; i < n; i++)
+	{
+		prev[i][i] = 1;
+	}
+	for (int a = 1; a < n; a++)
+	{
+		/*
+		 * dp[i][j] is the number of palindromes with length 2a + 1
+		 * starting on row i and ending on row j
+		 * with the middle of the string on the diagonal
+		 */
+		long long dp[500][500]{};
```

Comment:
would prefer if you put this outside `for (a = 1...)` bc this reallocates `dp` every iteration

Reply:
i, think that's intended?

Reply:
use memset or smth

Reply:
that's a bit too advanced tbh

Reply:
i think reallocating dp every time is fine, it doesn't run slower or anything

unless there's something bad that could happen this way? idk

Reply:
alr